src/hardware/vga_dac.cpp

```diff
@@ -127,35 +121,48 @@ static void vga_dac_send_color(const uint8_t palette_idx, const uint8_t color_id
 		LOG_TRACE("VGA: %s", msg.c_str());
 	}
 #endif
-
-	// If a palette entry was set to a non-EGA colour after the mode change
-	// was completed, we need to notify the renderer so it can re-init
-	// itself and potentially switch the current shader.
+	// We only want to trigger the "VGA DAC colours in EGA mode" detection
+	// logic when we're outside of a video mode change. Mode changes also
+	// set up the default CGA and EGA palette appropriate for the given
+	// mode, and that would only confuse and complicate the detection logic.
 	//
-	if (machine == MCH_VGA && !vga.ega_mode_with_vga_colors &&
-	    !vga.mode_change_in_progress &&
-	    bios_mode_number <= MaxEgaBiosModeNumber) {
-		bool non_ega_color = false;
-
-		if (is_640x350_16color_mode) {
-			// The 640x350 16-colour EGA mode (mode 10h) is special:
-			// the 16 colors can be freely chosen from a gamut of 64
-			// colours (6-bit RGB).
-			non_ega_color = !is_ega_color(rgb666);
-		} else {
-			// In all other EGA modes, the fixed "canonical
-			// 16-element CGA palette" (as emulated by VGA cards) is
-			// used.
-			non_ega_color = !is_cga_color(rgb666);
-		}
+	// In theory, if a program completely bypassed the INT 10h set video
+	// mode call and performed the mode change 100% itself by writing to the
+	// VGA registers directly, that would cause this logic not to trigger.
+	// Fortunately, no commercial game developers seemed to use such
+	// horrible practices.
+	//
+	if (machine == MCH_VGA && !vga.mode_change_in_progress &&
+	    !vga.ega_mode_with_vga_colors) {
+
+		const auto curr_mode = VGA_GetCurrentVideoMode();
+
+		const auto is_non_ega_color = [&]() {
+			if (curr_mode.bios_mode_number == ega_mode_640x350_16color) {
+				// The 640x350 16-colour EGA mode (mode 10h) is
+				// special: the 16 colors can be freely chosen
+				// from a gamut of 64 colours (6-bit RGB).
+				return !is_ega_color(rgb666);
+			} else {
+				// In all other EGA modes, the fixed "canonical
+				// 16-element CGA palette" (as emulated by VGA
+				// cards) is used.
+				return !is_cga_color(rgb666);
+			}
+		};
+
+		if (curr_mode.bios_mode_number <= MaxEgaBiosModeNumber &&
+		    is_non_ega_color()) {
 
-		if (non_ega_color) {
 			vga.ega_mode_with_vga_colors = true;
-#if 0
-			LOG_TRACE(
+
+			LOG_DEBUG(
 			        "VGA: EGA mode with VGA palette detected, "
 			        "notifying renderer");
-#endif
```

Comment:
Do we want this log message printed out to users?  Or does `LOG_DEBUG` get compiled out on a release build?  I can't remember.

Reply:
You only get `LOG_DEBUG` messages in debug builds, yeah. It's not noisy at all as the EGA mode with VGA palette is relatively rare, so I removed the `#if 0` from around it. Less cruft that way.